src/simstation/Agent.java

```diff
@@ -1,13 +1,122 @@
 package simstation;
 
-public class Agent {
+// Assuming Heading is an enum already defined somewhere
+import mvc.Model;
+import mvc.Utilities;
 
+import java.io.Serializable;
+
+public abstract class Agent implements Runnable, Serializable {
+
+    private String name;
     protected Heading heading;
-    public Agent() {
+    public int xc, yc; // Presumably for coordinates, but not shown in your diagram
+    private boolean suspended = false;
+    private boolean stopped = false;
+    private transient Thread myThread; // Transient because Threads are not serializable
+    public Simulation SimStation;
 
+    public Agent(Model model) {
+        this.heading = Heading.random(); // Assign a random heading
+        this.xc = Utilities.rng.nextInt()%400; // Example usage, needs proper range
+        this.yc = Utilities.rng.nextInt()%500; // Example usage, needs proper range
+        SimStation = (Simulation) model;
     }
 
-    protected void move(int steps) {
 
+
+    public String getName() {
+        return name;
+    }
+    private synchronized void checkSuspended() {
+        try {
+            while (!stopped && suspended) {
+                wait();
+            }
+        } catch (InterruptedException e) {
+            // Handle interruption
+        }
+    }
+    public synchronized void start() {
+        stopped = false;
+        suspended = false;
+        myThread = new Thread(this);
+        myThread.start();
+    }
+
+    public synchronized void suspend() {
+        suspended = true;
+    }
+
+    public synchronized void resume() {
+        if (suspended) {
+            suspended = false;
+            notify();
+        }
+    }
+
+    public synchronized void stop() {
+        stopped = true;
+        if (myThread != null) {
+            myThread.interrupt();
+        }
+    }
+
+    public synchronized boolean isSuspended() {
+        return suspended;
     }
+
+    public synchronized boolean isStopped() {
+        return stopped;
+    }
+
+    public synchronized void join() throws InterruptedException {
+        if (myThread != null) {
+            myThread.join();
+        }
+    }
+
+    @Override
+    public void run() {
+        myThread = Thread.currentThread();
+        while (!isStopped()) {
+            if (!isSuspended()) {
+                update();
+            }
+            try {
+                Thread.sleep(20); // Sleep time can be adjusted or made dynamic
+                checkSuspended();
+            } catch (InterruptedException e) {
+                // Check if the thread was interrupted to stop
+            }
+        }
+    }
+
+    public abstract void update();
+
+    public synchronized String toString() {
+        return name + " [" + heading + "] (" +
+                ((stopped) ? "stopped" : (suspended) ? "suspended" : "running") + ")";
+    }
+
+    public void move(int steps) {
+        Agent Neighbor = SimStation.getNeighbors(this,10);
```

Comment:
we need to pass in the model into when the move command is called. this can be passed through the run command all the way down to when we call this function